### ISSUE TITLE
Callbacks, save best only

### ISSUE THREAD

**Issue by Sat0ri** [keras-rl/keras-rl#162]
Hello,

how can I use **callbacks** and **save_best_only = True**?
or any other way to save best weights during **dqn.fit(...**.

Thanks

**Comment by mattdornfeld** [comment]
You would have to implement your own callback. You can't use the on that comes with Keras because keras_rl doesn't have epochs. The interface for the keras_rl Callback looks like so

```
class Callback(KerasCallback):
    def _set_env(self, env):
        self.env = env

    def on_episode_begin(self, episode, logs={}):
        pass

    def on_episode_end(self, episode, logs={}):
        pass

    def on_step_begin(self, step, logs={}):
        pass

    def on_step_end(self, step, logs={}):
        pass

    def on_action_begin(self, action, logs={}):
        pass

    def on_action_end(self, action, logs={}):
        pass
```
You would probably want to save your model at the end of each episode and keep the one with best total reward over all previous episodes. I imagine you could also extract some code from the Keras [ModelCheckPoint Callback](https://github.com/keras-team/keras/blob/master/keras/callbacks.py).

**Comment by RaphaelMeudec** [comment]
As by @mattdornfeld, best option would be to create a callback which holds the current best reward, and update it on episode_end or step_end. Then, at each episode_end (e.g), call a save on your weights if reward has been improved.

**Comment by guillem74** [comment]
@RaphaelMeudec @Sat0ri @mattdornfeld I've implemented that in my project, just take a look at:

https://github.com/guillem74/DRLDBackEnd/blob/master/callbacks.py#L179

Basically what I am doing is calling TrainEpisodeLogger callback when training. I modified the callback in order to compare after each episode if the reward is the highest at the moment, and if it is, I save the checkpoint weights. So basically, I have edited the on_episode_end() function by adding this:

        '''
        Code for saving up weights if the episode reward is higher than the last one
        '''
        
        if np.sum(self.rewards[episode])>self.lastreward:
            
            previousWeights = 'checkpoint_reward_{}.h5f'.format(self.lastreward)
            if os.path.exists(previousWeights): os.remove(previousWeights)
            self.lastreward = np.sum(self.rewards[episode])
            print("The reward is higher than the best one, saving checkpoint weights")
            newWeights = 'checkpoint_reward_{}.h5f'.format(np.sum(self.rewards[episode]))
            self.model.save_weights(newWeights, overwrite=True)
            
        else:
            print("The reward is lower than the best one, checkpoint weights not updated")
            
I also initialize self.lastreward = -200 in order to not save weights before -200 of episode reward.

Hope it helps you!
